server/services/v1/billing/reporter.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2022-2023 Tigris Data, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package billing
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/rs/zerolog/log"
+	"github.com/tigrisdata/tigris/errors"
+	"github.com/tigrisdata/tigris/server/config"
+	"github.com/tigrisdata/tigris/server/metadata"
+	"github.com/tigrisdata/tigris/server/metrics"
+	ulog "github.com/tigrisdata/tigris/util/log"
+)
+
+type UsageReporter struct {
+	glbState   metrics.UsageProvider
+	billingSvc Provider
+	interval   time.Duration
+	tenantMgr  metadata.NamespaceMetadataMgr
+	ctx        context.Context
+}
+
+func NewUsageReporter(gs metrics.UsageProvider, tm metadata.NamespaceMetadataMgr, billing Provider) (*UsageReporter, error) {
+	if gs == nil || tm == nil {
+		return nil, errors.Internal("usage reporter cannot be initialized")
+	}
+
+	return &UsageReporter{
+		glbState:   gs,
+		billingSvc: billing,
+		interval:   config.DefaultConfig.Billing.Reporter.RefreshInterval,
+		tenantMgr:  tm,
+		ctx:        context.Background(),
+	}, nil
+}
+
+func (r *UsageReporter) Start() {
+	if config.DefaultConfig.Billing.Reporter.Enabled {
+		go r.refreshLoop()
+	}
+}
+
+func (r *UsageReporter) refreshLoop() {
+	log.Info().Dur("refresh_interval", r.interval).Msg("Starting usage reporter")
+	t := time.NewTicker(r.interval)
+	defer t.Stop()
+	for range t.C {
+		ulog.E(r.push())
+	}
+}
+
+func (r *UsageReporter) push() error {
+	chunk := r.glbState.Flush()
+
+	if len(chunk.Tenants) == 0 {
+		log.Info().Msg("no tenant data to report")
+		return nil
+	}
+
+	trxnSuffix := strconv.FormatInt(chunk.EndTime.Unix(), 10)
+	events := make([]*UsageEvent, 0, len(chunk.Tenants))
+
+	for namespaceId, stats := range chunk.Tenants {
+		nsMeta := r.tenantMgr.GetNamespaceMetadata(r.ctx, namespaceId)
```

Comment:
Instead of this, you have an explicit method in tenantMgr that returns `accounts`, and that method will try to read from the tenant cache, and if it is empty, then try to read from the database. This will then avoid triggering the UpdateNamespaceMetadata below if the metadata is updated, but the worker view is stale.